activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,34 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def validation_needed?(klass, record, attribute)
+        return true if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        attributes = scope + [attribute.to_s]
+
+        return true if attributes.any? { |attribute| record.class._reflect_on_association(attribute) }
+        return true if attributes.any? { |attribute| record.attribute_changed?(attribute) ||
+                                                     record.read_attribute(attribute).nil? }
```

Comment:
Should this be done in a single iteration?
```suggestion
        return true if attributes.any? { |attribute| record.class._reflect_on_association(attribute) ||
                                                     record.attribute_changed?(attribute) ||
                                                     record.read_attribute(attribute).nil? }
```

Reply:
It can, but it looks more clear to me separately - one iteration checks for the changes and another if a reflection. Wdyt?

Reply:
It probably won't matter for performance, but to me with two iterations it would have me asking "Why are there two iterations? Is there some special case I'm missing?"
A seperate method that covers both cases like `attribute_validation_needed?` might be overkill. So 🤷 

I do think we should use a different name for the block variable as `attribute` mirrors the `attribute` passed into `validation_needed?`